### DISCUSSION TITLE
With graphql-java-spring-webflux, perform a non-blocking call to a backend service during Instrumentation

### DISCUSSION THREAD

**Discussion by CNAChino** [graphql-java/graphql-java#2408] [General]
Hi everyone, need help on this use case:

If one or more fields in SelectionSet is tagged with `@SomeDirective` in Schema, call a backend service to get more data/context about the request, then add to GraphQLContext.  This should be done before resolving fields.

Call to backend service must be non-blocking.  I'm using `graphql-java-spring-webflux`.

Tried `Instrumentation.beginExecution()` but was not able to find a way to have a non-blocking call.  Also tried blocking call just to experiment but Spring Webflux complaints that `.block()` is not allowed on the http thread pool.  

1. Is this a good use case for Instrumentation?
2. What's the recommended way to implement? 

Thank you in advance

**Comment by bbakerman** [comment]
A DataFetcher can return a CompleteableFuture (and async thing) and hence you can make all fetchers async

WebFlux allows you to turn a Mono into a CompleteableFuture easily.


Check out `graphql.execution.instrumentation.Instrumentation#instrumentDataFetcher` and `graphql.execution.DataFetcherResult`

this allows you to change how a data fetcher is implemented.  You get the OLD fetcher and you can make a new one

 and hence you can

```
    DataFetcher<?> instrumentDataFetcher(DataFetcher<?> oldDataFetcher, InstrumentationFieldFetchParameters parameters) {
    
    
       DataFetcher<?> newSecureDF = env -> {
            CompletableFuture cf = callSecurityLayer();
            return cf.thenApply( securityCall -> {
                 if (securityCall.failed()) {
                     return DataFetcherResult.error("Not secure");
                 }
                 return oldDataFetcher.get(parameters.getEnvironment());
            }
      };
    
       return newSecureDF;
```

**Reply by bbakerman** [reply]
See also https://www.graphql-java.com/blog/threads/
